modules/Bio/EnsEMBL/Variation/DBSQL/PublicationAdaptor.pm

```diff
@@ -226,7 +226,170 @@ sub fetch_all_by_Variation {
 
 }
 
+=head2 fetch_sources_by_pmid
 
+  Arg [1]    : String $pmid
+  Example    : $publication_adaptor->fetch_sources_by_pmid( 1234 );
+  Description: Retrieves a hashref of variants linked to publication sources via a pmid
+  Returntype : hashref
+  Exceptions : throw if pmid argument is not defined
+  Caller     : general
+  Status     : At Risk
+
+=cut
+
+sub fetch_sources_by_pmid {
+  my $self = shift;
+  my $pmid = shift;
+
+  throw("pmid must be defined") if(!defined($pmid));
+
+  my $pub = $self->fetch_by_pmid($pmid);
+  my $sources = $self->fetch_sources($pub->{dbID});
+
+  return $sources
+}
+
+=head2 fetch_sources_by_pmcid
+
+  Arg [1]    : String $pmcid
+  Example    : $publication_adaptor->fetch_sources_by_pmcid( 'PMC1234' );
+  Description: Retrieves a hashref of variants linked to publication sources via a pmcid
+  Returntype : hashref
+  Exceptions : throw if pmcid argument is not defined
+  Caller     : general
+  Status     : At Risk
+
+=cut
+
+sub fetch_sources_by_pmcid {
+  my $self = shift;
+  my $pmcid = shift;
+
+  throw("pmcid must be defined") if(!defined($pmcid));
+
+  my $pub = $self->fetch_by_pmcid($pmcid);
+  my $sources = $self->fetch_sources($pub->{dbID});
+
+  return $sources;
+}
+
+=head2 fetch_sources
+
+  Arg [1]    : String $publication_id
+  Example    : $publication_adaptor->fetch_sources( $publication_id );
+  Description: Retrieves a hashref of variants linked to publication sources via a publication_id
+  Returntype : hashref
+  Exceptions : throw if variation is not defined
+               throw if attribute is not defined
+  Caller     : general
+  Status     : At Risk
+
+=cut
+
+sub fetch_sources {
+  my $self   = shift;
+  my $pub_id = shift;
+
+  my $sth = $self->prepare(qq{SELECT variation_id,data_source_attrib FROM variation_citation WHERE publication_id = ? });
+  $sth->execute($pub_id);
+  my $data = $sth->fetchall_arrayref();
+
+  my %result;
+
+  foreach my $var_source (@{$data}) {
```

Comment:
You can first collect all variation ids and source attrib ids and then create only one query for each. For example:
SELECT variation_id, name FROM variation WHERE variation_id IN (id1, id2, id3). You can share the code for getting attributes in a separate method because it is used again in fetch_sources_by_variation.